File: critic/extractor.py
"""
Extractor - extracts structured information from each chunk.

This is the detailed extraction pass that runs after discovery.
Uses the Message Batches API for 50% cost reduction.
"""

import json
import re
import time
from typing import Any, Callable, Optional, TYPE_CHECKING

import anthropic

from critic.types import Chunk
from critic.schema import (
    ChunkWithText,
    ChunkExtraction,
    EventExtraction,
    CharacterMention,
    FactExtraction,
    PlotThreadTouch,
    SetupExtraction,
    DialogueLine,
    SceneBreak,
    TextLocation,
)
from critic.discovery import DiscoveredEntities, format_discovered_context

if TYPE_CHECKING:
    from critic.cache import AnalysisCache

from critic.config import DEFAULT_MODEL

# Polling interval for batch status (seconds)
BATCH_POLL_INTERVAL = 5


def build_extraction_prompt(discovered_context: Optional[str] = None) -> str:
    """Build the extraction prompt, optionally with discovered entity context."""
    context_section = ""
    if discovered_context:
        context_section = f"""
# DOCUMENT CONTEXT
The following characters and plot threads have been identified in this manuscript. Use these as reference when extracting:

{discovered_context}

---
"""

    character_instructions = """For each character that appears or is mentioned:
- name: Character's name (use the canonical name from the known characters list if they match)
- role: "present" (in the scene), "mentioned" (talked about), or "flashback"
- attributes: Array of character details. For each attribute:
  - value: The attribute (e.g., "blue eyes", "detective", "nervous")
  - category: One of:
    - "physical": Stable physical traits (eye color, height, scars, age)
    - "personality": Stable personality traits (determined, kind, suspicious)
    - "occupation": Job or role (detective, fisherman, bartender)
    - "action": Momentary actions (clutching coffee, leaning forward) - SKIP THESE, don't include actions
  IMPORTANT: Only include stable traits (physical, personality, occupation). Skip momentary actions.
- relationships: Array of relationships FROM THIS CHARACTER'S PERSPECTIVE. Format: {target: "other character name", relationship: "what the target is TO this character"}
  IMPORTANT: The relationship describes what the TARGET is to the SOURCE character.
  Examples: If Alice's mother is named Beth, Alice's relationship entry is {target: "Beth", relationship: "mother"} (Beth is Alice's mother).
- quote: A quote showing their appearance/mention"""

    thread_instructions = """Story threads touched in this section. Match to known threads when possible. A thread is resolved when its central question/conflict is answered/concluded.
- name: Use the exact name from the known plot threads list if this section touches that thread
- action: "introduced" (new thread), "advanced" (progress made), "complicated" (new obstacle/twist), or "resolved" (concluded)
- description: What happened with this thread in this section
- quote: Relevant quote"""

    return f"""You are analyzing a section of a novel manuscript. Extract structured information with EXACT QUOTES from the text.
{context_section}
For each item you extract, you MUST include:
1. The exact quote from the text (verbatim, word-for-word)
2. This quote will be used to locate the text in the original document, so accuracy is critical

Analyze the text and extract:

## 1. Events (things that happen)
For each significant event:
- description: What happened
- timeMarker: Quote ANY time reference from the text. Look for:
  * Explicit times: "Monday morning", "3pm", "January 5th"
  * Relative times: "three days later", "the next morning", "after dinner"
  * Vague times: "that evening", "later", "soon after"
  * Scene context: "during the meeting", "while driving home"
  * If truly no time reference, use "continuing" (same scene as previous) or "scene break" (new scene, time unclear)
- precision: "exact" (specific date/time), "relative" (e.g., "three days later"), or "vague" (e.g., "that morning")
- sequenceNote: What narrative context surrounds this?
- characters: Names of characters involved
- quote: The EXACT sentence(s) describing this event

## 2. Character Appearances
{character_instructions}

## 3. Facts Established
Concrete details that could be contradicted later:
- content: The fact
- category: "character", "world", "location", "object", "relationship", or "other"
- subject: What/who it's about
- quote: EXACT quote establishing this fact

## 4. Plot Threads
{thread_instructions}

## 5. Setups/Foreshadowing
Things that seem to promise future payoff:
- description: What was set up
- weight: "subtle", "moderate", or "heavy"
- impliedPayoff: What this seems to promise
- quote: The setup text

## 6. Dialogue (important conversations only)
For significant dialogue exchanges that reveal character or advance plot:
- speaker: Who is speaking
- target: Who they're speaking to (if clear)
- summary: Brief summary of what they said (2-3 sentences max)
- tone: The emotional tone (e.g., "angry", "pleading", "casual", "threatening")
- reveals: Array of key information revealed in this dialogue
- quote: A key quote from this dialogue

## 7. Scene Structure
When the scene changes (location, time, or POV shift):
- sceneNumber: Number starting from 1 within this chapter
- location: Where the scene takes place
- time: Time reference for this scene (if mentioned)
- charactersPresent: Array of characters in the scene
- povCharacter: Whose perspective we're seeing (if clear)

## 8. Open Questions
Mysteries or questions raised for the reader (just list as strings)

Return as JSON:
{{
  "events": [...],
  "characters": [...],
  "facts": [...],
  "plotThreads": [...],
  "setups": [...],
  "dialogue": [...],
  "scenes": [...],
  "openQuestions": [...]
}}

CRITICAL: Every quote must be EXACTLY as it appears in the text. Do not paraphrase."""


def extract_from_chunks(
    chunks: list[Chunk],
    api_key: Optional[str] = None,
    model: str = DEFAULT_MODEL,
    discovered_entities: Optional[DiscoveredEntities] = None,
    cache: Optional["AnalysisCache"] = None,
    on_progress: Optional[Callable[[int, int, str], None]] = None,
) -> tuple[list[ChunkWithText], dict]:
    """
    Extract structured data from chunks using the Message Batches API.

    Uses batch processing for 50% cost reduction.

    Args:
        chunks: Document chunks to extract from
        api_key: Anthropic API key
        model: Model to use
        discovered_entities: Pre-discovered entities for context
        cache: Optional cache for extraction results
        on_progress: Callback (completed, total, chunk_id)

    Returns:
        Tuple of (chunks_with_text, token_usage)
    """
    client = anthropic.Anthropic(api_key=api_key)

    # Build prompt with discovered context
    discovered_context = (
        format_discovered_context(discovered_entities)
        if discovered_entities else None
    )
    extraction_prompt = build_extraction_prompt(discovered_context)

    # Hash for cache key
    entities_hash = ""
    if discovered_entities and cache:
        entities_hash = cache.hash_discovered_entities(discovered_entities)

    results: list[Optional[ChunkWithText]] = [None] * len(chunks)
    token_usage = {'input_tokens': 0, 'output_tokens': 0}
    completed = 0

    # Calculate offsets
    current_offset = 0
    offsets = []
    for chunk in chunks:
        offsets.append(current_offset)
        current_offset += len(chunk.content) + 1

    # Check cache for each chunk
    uncached_indices = []
    for i, chunk in enumerate(chunks):
        if cache:
            cached = cache.get_extraction(chunk.content, chunk.id, model, entities_hash)
            if cached:
                extraction, usage = cached
                results[i] = ChunkWithText(
                    id=chunk.id,
                    title=chunk.title,
                    text=chunk.content,
                    start_offset=offsets[i],
                    end_offset=offsets[i] + len(chunk.content),
                    extraction=extraction,
                )
                token_usage['input_tokens'] += usage['input_tokens']
                token_usage['output_tokens'] += usage['output_tokens']
                completed += 1
                if on_progress:
                    on_progress(completed, len(chunks), chunk.id)
                continue
        uncached_indices.append(i)

    if not uncached_indices:
        return [r for r in results if r is not None], token_usage

    print(f"[Extractor] Processing {len(uncached_indices)} uncached chunks "
          f"({len(chunks) - len(uncached_indices)} from cache) via batch API")

    # Build batch requests
    batch_requests = []
    for idx in uncached_indices:
        chunk = chunks[idx]
        chunk_context = (
            f"## Section: {chunk.title}\n\n{chunk.content}"
            if chunk.title else chunk.content
        )
        batch_requests.append({
            "custom_id": f"chunk-{idx}",
            "params": {
                "model": model,
                "max_tokens": 64000,
                "messages": [{
                    "role": "user",
                    "content": f"{extraction_prompt}\n\n---\n\nTEXT TO ANALYZE:\n\n{chunk_context}",
                }],
            },
        })

    # Submit batch
    batch = client.messages.batches.create(requests=batch_requests)
    print(f"[Extractor] Batch submitted: {batch.id}")

    # Poll for completion
    while batch.processing_status != "ended":
        time.sleep(BATCH_POLL_INTERVAL)
        batch = client.messages.batches.retrieve(batch.id)

        # Count completed
        batch_completed = (batch.request_counts.succeeded +
                          batch.request_counts.errored +
                          batch.request_counts.canceled +
                          batch.request_counts.expired)
        total_in_batch = len(uncached_indices)

        if on_progress:
            on_progress(completed + batch_completed, len(chunks), f"batch:{batch.id[:8]}")

        print(f"[Extractor] Batch progress: {batch_completed}/{total_in_batch} "
              f"(status: {batch.processing_status})")

    # Process results
    print(f"[Extractor] Batch complete. Succeeded: {batch.request_counts.succeeded}, "
          f"Errored: {batch.request_counts.errored}")

    for result in client.messages.batches.results(batch.id):
        # Parse chunk index from custom_id
        idx = int(result.custom_id.split("-")[1])
        chunk = chunks[idx]

        if result.result.type == "succeeded":
            message = result.result.message
            text = message.content[0].text if message.content else ""

            usage = {
                'input_tokens': message.usage.input_tokens,
                'output_tokens': message.usage.output_tokens,
            }
            token_usage['input_tokens'] += usage['input_tokens']
            token_usage['output_tokens'] += usage['output_tokens']

            raw = _parse_json_response(text, chunk.id)
            extraction = _process_raw_extraction(raw, chunk)

            results[idx] = ChunkWithText(
                id=chunk.id,
                title=chunk.title,
                text=chunk.content,
                start_offset=offsets[idx],
                end_offset=offsets[idx] + len(chunk.content),
                extraction=extraction,
            )

            # Save to cache
            if cache:
                cache.set_extraction(
                    chunk.content, model, entities_hash,
                    extraction, usage
                )
        else:
            # Handle error - create empty extraction
            print(f"[Extractor] Error for chunk {chunk.id}: {result.result.type}")
            if hasattr(result.result, 'error'):
                print(f"  Error details: {result.result.error}")

            results[idx] = ChunkWithText(
                id=chunk.id,
                title=chunk.title,
                text=chunk.content,
                start_offset=offsets[idx],
                end_offset=offsets[idx] + len(chunk.content),
                extraction=ChunkExtraction(),
            )

        completed += 1

    return [r for r in results if r is not None], token_usage


def _normalize_attribute(a: Any) -> Optional[dict]:
    """Normalize an attribute to {value, category} dict. Returns None for actions."""
    if isinstance(a, str):
        # Legacy format - treat as unknown category
        return {"value": a, "category": "state"}
    if isinstance(a, dict):
        category = a.get('category', 'state')
        # Skip actions - they're momentary, not stable traits
        if category == 'action':
            return None
        value = a.get('value') or a.get('attribute') or a.get('name', '')
        if value:
            return {"value": value, "category": category}
    return None


def _process_raw_extraction(raw: dict, chunk: Chunk) -> ChunkExtraction:
    """Process raw extraction and add text locations."""

    def loc(quote: str) -> TextLocation:
        return _find_text_location(quote, chunk.content, chunk.id)

    def make_id(prefix: str, i: int) -> str:
        return f"{chunk.id}-{prefix}-{i}"

    # Process events - store character names (will be resolved to IDs in aggregation)
    events = [
        EventExtraction(
            id=make_id("event", i),
            description=e.get('description', ''),
            time_marker=e.get('timeMarker', ''),
            precision=e.get('precision', 'vague'),
            sequence_note=e.get('sequenceNote'),
            character_ids=e.get('characters', []),  # Model-provided character names
            location=loc(e.get('quote', '')),
        )
        for i, e in enumerate(raw.get('events', []))
    ]

    # Process characters (needs special handling for attributes)
    character_mentions = []
    for c in raw.get('characters', []):
        # Normalize attributes - filter out actions, keep stable traits
        attrs = [a for a in (_normalize_attribute(x) for x in c.get('attributes', [])) if a]
        character_mentions.append(CharacterMention(
            character_id='',
            name=c.get('name', ''),
            role=c.get('role', 'present'),
            location=loc(c.get('quote', '')),
            # Store as list of dicts with value and category
            attributes_mentioned=attrs,
            relationships_mentioned=[
                {'target': r.get('target', ''), 'relationship': r.get('relationship', '')}
                for r in c.get('relationships', [])
            ],
        ))

    # Process facts
    facts = [
        FactExtraction(
            id=make_id("fact", i),
            content=f.get('content', ''),
            category=f.get('category', 'other'),
            subject=f.get('subject', ''),
            location=loc(f.get('quote', '')),
        )
        for i, f in enumerate(raw.get('facts', []))
    ]

    # Process plot threads
    plot_threads = [
        PlotThreadTouch(
            thread_id='',
            name=pt.get('name', ''),
            action=pt.get('action', 'advanced'),
            description=pt.get('description', ''),
            location=loc(pt.get('quote', '')),
        )
        for pt in raw.get('plotThreads', [])
    ]

    # Process setups
    setups = [
        SetupExtraction(
            id=make_id("setup", i),
            description=s.get('description', ''),
            weight=s.get('weight', 'moderate'),
            implied_payoff=s.get('impliedPayoff', ''),
            location=loc(s.get('quote', '')),
            status='pending',
        )
        for i, s in enumerate(raw.get('setups', []))
    ]

    # Process dialogue
    dialogue = [
        DialogueLine(
            speaker=d.get('speaker', ''),
            target=d.get('target'),
            summary=d.get('summary', ''),
            tone=d.get('tone'),
            reveals=d.get('reveals', []),
            location=loc(d.get('quote', '')),
        )
        for d in raw.get('dialogue', [])
    ]

    # Process scenes
    scenes = []
    for s in raw.get('scenes', []):
        scene_loc = loc(s.get('location', '') or s.get('time', ''))
        scenes.append(SceneBreak(
            scene_number=s.get('sceneNumber', len(scenes) + 1),
            location=s.get('location'),
            time=s.get('time'),
            characters_present=s.get('charactersPresent', []),
            pov_character=s.get('povCharacter'),
            start_offset=scene_loc.start_offset,
            end_offset=scene_loc.end_offset,
        ))

    return ChunkExtraction(
        events=events,
        character_mentions=character_mentions,
        facts=facts,
        plot_threads=plot_threads,
        setups=setups,
        dialogue=dialogue,
        scenes=scenes,
        open_questions=raw.get('openQuestions', []),
    )


def _find_text_location(quote: str, chunk_text: str, chunk_id: str) -> TextLocation:
    """Find the location of a quote in the chunk text."""
    # Try exact match
    start = chunk_text.find(quote)

    # Try case-insensitive
    if start == -1:
        start = chunk_text.lower().find(quote.lower())

    # Try fuzzy match
    if start == -1:
        start = _fuzzy_find_quote(quote, chunk_text)

    if start != -1:
        end = start + len(quote)
        return TextLocation(
            chunk_id=chunk_id,
            start_offset=start,
            end_offset=end,
            snippet=chunk_text[start:end],
            human_readable=_get_human_readable_location(chunk_text, start),
        )

    # Fallback
    return TextLocation(
        chunk_id=chunk_id,
        start_offset=0,
        end_offset=min(len(quote), len(chunk_text)),
        snippet=quote,
        human_readable="Location approximate",
    )


def _fuzzy_find_quote(quote: str, text: str) -> int:
    """Fuzzy find a quote in text."""
    def normalize(s: str) -> str:
        return re.sub(r'[^\w\s]', '', s.lower()).strip()

    normalized_quote = normalize(quote)
    normalized_text = normalize(text)

    # Try normalized version
    idx = normalized_text.find(normalized_quote)
    if idx != -1:
        ratio = idx / len(normalized_text) if normalized_text else 0
        return int(ratio * len(text))

    # Try first few words
    words = normalized_quote.split()[:5]
    if len(words) >= 3:
        partial = ' '.join(words)
        idx = normalized_text.find(partial)
        if idx != -1:
            ratio = idx / len(normalized_text) if normalized_text else 0
            return int(ratio * len(text))

    return -1


def _get_human_readable_location(text: str, offset: int) -> str:
    """Get human-readable location."""
    before = text[:offset]
    paragraphs = before.split('\n\n')
    return f"Paragraph {len(paragraphs)}"


def _parse_json_response(text: str, chunk_id: str = "") -> dict:
    """Parse JSON from Claude's response."""
    # Try code block
    code_match = re.search(r'```(?:json)?\s*([\s\S]*?)```', text)
    if code_match:
        try:
            return json.loads(code_match.group(1).strip())
        except json.JSONDecodeError as e:
            print(f"JSON parse error in code block for {chunk_id}: {e}")

    # Try raw JSON
    json_match = re.search(r'\{[\s\S]*\}', text)
    if json_match:
        try:
            return json.loads(json_match.group(0))
        except json.JSONDecodeError as e:
            print(f"JSON parse error for {chunk_id}: {e}")
            print(f"JSON snippet: {json_match.group(0)[:200]}...")

    print(f"Failed to extract JSON from response for {chunk_id}")
    return {}

File: app/core/src/extractor.ts
/**
 * Extractor - Captures source text locations for frontend highlighting
 *
 * This is the second pass of extraction, after discovery has identified
 * all characters and plot threads in the document.
 */

import Anthropic from '@anthropic-ai/sdk';
import type { Chunk } from './types.js';
import type {
  ChunkWithText,
  ChunkExtraction,
  EventExtraction,
  CharacterMention,
  FactExtraction,
  PlotThreadTouch,
  SetupExtraction,
  TextLocation,
} from './output-schema.js';
import type { DiscoveredEntities } from './discovery.js';
import { formatDiscoveredContext } from './discovery.js';

import { AnalysisCache, hashDiscoveredEntities } from './cache.js';

export interface ExtractorOptions {
  apiKey?: string;
  model?: string;
  concurrency?: number;
  /** Pre-discovered entities to provide as context */
  discoveredEntities?: DiscoveredEntities;
  /** Cache instance for storing/retrieving extraction results */
  cache?: AnalysisCache;
  onProgress?: (completed: number, total: number, chunkId: string) => void;
}

const DEFAULT_MODEL = 'claude-sonnet-4-20250514';

/**
 * Build the extraction prompt, optionally with discovered entity context
 */
function buildExtractionPrompt(discoveredContext?: string): string {
  const contextSection = discoveredContext
    ? `\n# DOCUMENT CONTEXT\nThe following characters and plot threads have been identified in this manuscript. Use these as reference when extracting:\n\n${discoveredContext}\n\n---\n`
    : '';

  const characterInstructions = discoveredContext
    ? `For each character that appears or is mentioned (use the known character names when matching):
- name: Character's name (use the canonical name from the known characters list if they match)
- role: "present" (in the scene), "mentioned" (talked about), or "flashback"
- attributes: Array of physical/personality details, each as a string like "blue eyes", "tall", "nervous demeanor"
- relationships: Array of relationships FROM THIS CHARACTER'S PERSPECTIVE revealed in this section. Format: {target: "other character name", relationship: "what the target is TO this character"}
  IMPORTANT: The relationship describes what the TARGET is to the SOURCE character.
  Examples: If Alice's mother is named Beth, Alice's relationship entry is {target: "Beth", relationship: "mother"} (Beth is Alice's mother).
  If Bob employs Carol, Carol's relationship entry is {target: "Bob", relationship: "employer"} (Bob is Carol's employer).
- quote: A quote showing their appearance/mention`
    : `For each character that appears or is mentioned:
- name: Character's name
- role: "present" (in the scene), "mentioned" (talked about), or "flashback"
- attributes: Array of physical/personality details, each as a string like "blue eyes", "tall", "nervous demeanor"
- relationships: Array of relationships FROM THIS CHARACTER'S PERSPECTIVE revealed in this section. Format: {target: "other character name", relationship: "what the target is TO this character"}
  IMPORTANT: The relationship describes what the TARGET is to the SOURCE character.
  Examples: If Alice's mother is named Beth, Alice's relationship entry is {target: "Beth", relationship: "mother"} (Beth is Alice's mother).
  If Bob employs Carol, Carol's relationship entry is {target: "Bob", relationship: "employer"} (Bob is Carol's employer).
- quote: A quote showing their appearance/mention`;

  const threadInstructions = discoveredContext
    ? `Story threads touched in this section. Match to known threads when possible. A thread is resolved when its central question/conflict is answered/concluded.
- name: Use the exact name from the known plot threads list if this section touches that thread
- action: "introduced" (new thread), "advanced" (progress made), "complicated" (new obstacle/twist), or "resolved" (concluded - the central question is answered)
- description: What happened with this thread in this section
- quote: Relevant quote`
    : `Story threads touched in this section. A thread is resolved when its central question/conflict is answered/concluded (e.g., a mystery is solved, a goal is achieved, a relationship reaches a conclusion).
- name: Brief name for the thread (e.g., "Emma Hartley murder investigation", "Sarah and Marcus romance")
- action: "introduced" (new thread), "advanced" (progress made), "complicated" (new obstacle/twist), or "resolved" (concluded)
- description: What happened with this thread
- quote: Relevant quote`;

  return `You are analyzing a section of a novel manuscript. Extract structured information with EXACT QUOTES from the text.
${contextSection}
For each item you extract, you MUST include:
1. The exact quote from the text (verbatim, word-for-word)
2. This quote will be used to locate the text in the original document, so accuracy is critical

Analyze the text and extract:

## 1. Events (things that happen)
For each significant event:
- description: What happened
- timeMarker: Quote ANY time reference from the text. Look for:
  * Explicit times: "Monday morning", "3pm", "January 5th"
  * Relative times: "three days later", "the next morning", "after dinner"
  * Vague times: "that evening", "later", "soon after"
  * Scene context: "during the meeting", "while driving home"
  * If truly no time reference, use "continuing" (same scene as previous) or "scene break" (new scene, time unclear)
- precision: "exact" (specific date/time), "relative" (e.g., "three days later"), or "vague" (e.g., "that morning")
- sequenceNote: What narrative context surrounds this? (e.g., "after the argument", "before discovering the body", "same conversation as previous event")
- characters: Names of characters involved (use canonical names from known characters if applicable)
- quote: The EXACT sentence(s) describing this event

## 2. Character Appearances
${characterInstructions}

## 3. Facts Established
Concrete details that could be contradicted later:
- content: The fact
- category: "character", "world", "location", "object", "relationship", or "other"
- subject: What/who it's about
- quote: EXACT quote establishing this fact

## 4. Plot Threads
${threadInstructions}

## 5. Setups/Foreshadowing
Things that seem to promise future payoff:
- description: What was set up
- weight: "subtle", "moderate", or "heavy"
- impliedPayoff: What this seems to promise
- quote: The setup text

## 6. Open Questions
Mysteries or questions raised for the reader (just list as strings)

Return as JSON:
{
  "events": [...],
  "characters": [...],
  "facts": [...],
  "plotThreads": [...],
  "setups": [...],
  "openQuestions": [...]
}

CRITICAL: Every quote must be EXACTLY as it appears in the text. Do not paraphrase.`;
}

export interface ExtractionResult {
  chunks: ChunkWithText[];
  tokenUsage: {
    inputTokens: number;
    outputTokens: number;
  };
}

interface ChunkExtractionResult {
  chunk: ChunkWithText;
  inputTokens: number;
  outputTokens: number;
}

interface RawExtraction {
  events?: Array<{
    description: string;
    timeMarker: string;
    precision: 'exact' | 'relative' | 'vague';
    sequenceNote?: string;
    characters: string[];
    quote: string;
  }>;
  characters?: Array<{
    name: string;
    role: 'present' | 'mentioned' | 'flashback';
    attributes?: Array<string | Record<string, unknown>>;
    relationships?: Array<{
      target: string;
      relationship: string;
    }>;
    quote: string;
  }>;
  facts?: Array<{
    content: string;
    category: string;
    subject: string;
    quote: string;
  }>;
  plotThreads?: Array<{
    name: string;
    action: 'introduced' | 'advanced' | 'complicated' | 'resolved';
    description: string;
    quote: string;
  }>;
  setups?: Array<{
    description: string;
    weight: 'subtle' | 'moderate' | 'heavy';
    impliedPayoff: string;
    quote: string;
  }>;
  openQuestions?: string[];
}

/**
 * Extract structured data from chunks with source locations
 */
export async function extractFromChunks(
  chunks: Chunk[],
  options: ExtractorOptions = {}
): Promise<ExtractionResult> {
  const apiKey = options.apiKey || process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    throw new Error('ANTHROPIC_API_KEY is required');
  }

  const client = new Anthropic({ apiKey });
  const model = options.model || DEFAULT_MODEL;
  const concurrency = options.concurrency || 3;
  const cache = options.cache;

  // Build extraction prompt with discovered context if available
  const discoveredContext = options.discoveredEntities
    ? formatDiscoveredContext(options.discoveredEntities)
    : undefined;
  const extractionPrompt = buildExtractionPrompt(discoveredContext);

  // Hash of discovered entities for cache key
  const entitiesHash = options.discoveredEntities
    ? hashDiscoveredEntities(options.discoveredEntities)
    : '';

  const allResults: ChunkExtractionResult[] = [];
  let completed = 0;

  // Track character offset across chunks
  let currentOffset = 0;

  // Separate chunks into cached and uncached
  const uncachedChunks: Array<{ chunk: Chunk; offset: number; index: number }> = [];
  const chunkOffsets: number[] = [];

  for (let i = 0; i < chunks.length; i++) {
    const chunk = chunks[i];
    const offset = currentOffset;
    currentOffset += chunk.content.length + 1;
    chunkOffsets.push(offset);

    // Check cache
    if (cache) {
      const cached = cache.getExtraction(chunk.content, chunk.id, model, entitiesHash);
      if (cached) {
        allResults[i] = {
          chunk: {
            id: chunk.id,
            title: chunk.title,
            text: chunk.content,
            startOffset: offset,
            endOffset: offset + chunk.content.length,
            extraction: cached.extraction,
          },
          inputTokens: cached.tokenUsage.inputTokens,
          outputTokens: cached.tokenUsage.outputTokens,
        };
        completed++;
        options.onProgress?.(completed, chunks.length, chunk.id);
        continue;
      }
    }

    uncachedChunks.push({ chunk, offset, index: i });
  }

  if (uncachedChunks.length > 0) {
    console.log(`[Extractor] Processing ${uncachedChunks.length} uncached chunks (${chunks.length - uncachedChunks.length} from cache)`);
  }

  // Process uncached chunks in batches
  for (let i = 0; i < uncachedChunks.length; i += concurrency) {
    const batch = uncachedChunks.slice(i, i + concurrency);

    const batchResults = await Promise.all(
      batch.map(async ({ chunk, offset, index }) => {
        const result = await extractFromChunk(
          client,
          model,
          chunk,
          offset,
          extractionPrompt
        );

        // Save to cache
        if (cache) {
          cache.setExtraction(
            chunk.content,
            model,
            entitiesHash,
            result.chunk.extraction,
            { inputTokens: result.inputTokens, outputTokens: result.outputTokens }
          );
        }

        completed++;
        options.onProgress?.(completed, chunks.length, chunk.id);
        return { result, index };
      })
    );

    // Insert results at correct positions
    for (const { result, index } of batchResults) {
      allResults[index] = result;
    }
  }

  // Aggregate token usage
  const tokenUsage = {
    inputTokens: allResults.reduce((sum, r) => sum + r.inputTokens, 0),
    outputTokens: allResults.reduce((sum, r) => sum + r.outputTokens, 0),
  };

  return {
    chunks: allResults.map(r => r.chunk),
    tokenUsage,
  };
}

/**
 * Extract from a single chunk with location tracking
 */
async function extractFromChunk(
  client: Anthropic,
  model: string,
  chunk: Chunk,
  startOffset: number,
  extractionPrompt: string
): Promise<ChunkExtractionResult> {
  const chunkContext = chunk.title
    ? `## Section: ${chunk.title}\n\n${chunk.content}`
    : chunk.content;

  // Use streaming for long requests
  const stream = client.messages.stream({
    model,
    max_tokens: 64000,
    messages: [
      {
        role: 'user',
        content: `${extractionPrompt}\n\n---\n\nTEXT TO ANALYZE:\n\n${chunkContext}`,
      },
    ],
  });

  const response = await stream.finalMessage();

  // Check if response was truncated
  if (response.stop_reason === 'max_tokens') {
    console.warn(`Warning: Extraction for chunk ${chunk.id} was truncated - response hit token limit`);
  }

  const text = response.content
    .filter((block): block is Anthropic.TextBlock => block.type === 'text')
    .map((block) => block.text)
    .join('');

  const raw = parseJsonFromResponse(text, chunk.id) as RawExtraction;

  // Convert raw extraction to typed extraction with locations
  const extraction = processRawExtraction(raw, chunk);

  return {
    chunk: {
      id: chunk.id,
      title: chunk.title,
      text: chunk.content,
      startOffset,
      endOffset: startOffset + chunk.content.length,
      extraction,
    },
    inputTokens: response.usage.input_tokens,
    outputTokens: response.usage.output_tokens,
  };
}

/**
 * Process raw extraction and add text locations
 */
function processRawExtraction(
  raw: RawExtraction,
  chunk: Chunk
): ChunkExtraction {
  const findLocation = (quote: string): TextLocation => {
    return findTextLocation(quote, chunk.content, chunk.id);
  };

  // Process events
  const events: EventExtraction[] = (raw.events ?? []).map((e, idx) => ({
    id: `${chunk.id}-event-${idx}`,
    description: e.description,
    timeMarker: e.timeMarker,
    sequenceNote: e.sequenceNote,
    precision: e.precision,
    characterIds: [], // Will be linked in aggregation pass
    location: findLocation(e.quote),
  }));

  // Process characters
  const characterMentions: CharacterMention[] = (raw.characters ?? []).map((c) => ({
    characterId: '', // Will be assigned in aggregation pass
    name: c.name,
    role: c.role,
    location: findLocation(c.quote),
    attributesMentioned: (c.attributes ?? []).map((a) => {
      // Handle various formats Claude might return
      if (typeof a === 'string') {
        return a; // Already a string like "blue eyes"
      }
      if (a && typeof a === 'object') {
        const obj = a as Record<string, unknown>;
        // Try different field name combinations
        const attr = String(obj.attribute || obj.name || obj.trait || obj.type || '');
        const val = String(obj.value || obj.detail || obj.description || '');
        if (attr && val) {
          return `${attr}: ${val}`;
        }
        // If only one field exists, use it as the full description
        return attr || val || JSON.stringify(a);
      }
      return String(a);
    }).filter(Boolean),
    relationshipsMentioned: (c.relationships ?? []).map((r) => ({
      target: r.target,
      relationship: r.relationship,
    })),
  }));

  // Process facts
  const facts: FactExtraction[] = (raw.facts ?? []).map((f, idx) => ({
    id: `${chunk.id}-fact-${idx}`,
    content: f.content,
    category: f.category as FactExtraction['category'],
    subject: f.subject,
    location: findLocation(f.quote),
  }));

  // Process plot threads
  const plotThreads: PlotThreadTouch[] = (raw.plotThreads ?? []).map((pt) => ({
    threadId: '', // Will be assigned in aggregation pass
    name: pt.name,
    action: pt.action,
    description: pt.description,
    location: findLocation(pt.quote),
  }));

  // Process setups
  const setups: SetupExtraction[] = (raw.setups ?? []).map((s, idx) => ({
    id: `${chunk.id}-setup-${idx}`,
    description: s.description,
    weight: s.weight,
    impliedPayoff: s.impliedPayoff,
    location: findLocation(s.quote),
    status: 'pending' as const,
  }));

  return {
    events,
    characterMentions,
    facts,
    plotThreads,
    setups,
    openQuestions: raw.openQuestions ?? [],
  };
}

/**
 * Find the location of a quote in the chunk text
 */
function findTextLocation(
  quote: string,
  chunkText: string,
  chunkId: string
): TextLocation {
  // Try exact match first
  let startOffset = chunkText.indexOf(quote);

  // If not found, try case-insensitive
  if (startOffset === -1) {
    startOffset = chunkText.toLowerCase().indexOf(quote.toLowerCase());
  }

  // If still not found, try fuzzy match (find best substring match)
  if (startOffset === -1) {
    startOffset = fuzzyFindQuote(quote, chunkText);
  }

  // If we found it
  if (startOffset !== -1) {
    const endOffset = startOffset + quote.length;
    return {
      chunkId,
      startOffset,
      endOffset,
      snippet: chunkText.slice(startOffset, endOffset),
      humanReadable: getHumanReadableLocation(chunkText, startOffset),
    };
  }

  // Fallback: return beginning of chunk with the quote as snippet
  return {
    chunkId,
    startOffset: 0,
    endOffset: Math.min(quote.length, chunkText.length),
    snippet: quote,
    humanReadable: 'Location approximate',
  };
}

/**
 * Fuzzy find a quote in text (handles minor transcription differences)
 */
function fuzzyFindQuote(quote: string, text: string): number {
  // Normalize both strings
  const normalizeForSearch = (s: string) =>
    s.toLowerCase().replace(/[^\w\s]/g, '').replace(/\s+/g, ' ').trim();

  const normalizedQuote = normalizeForSearch(quote);
  const normalizedText = normalizeForSearch(text);

  // Try to find normalized version
  const normalizedIdx = normalizedText.indexOf(normalizedQuote);
  if (normalizedIdx !== -1) {
    // Map back to original text position (approximate)
    const ratio = normalizedIdx / normalizedText.length;
    return Math.floor(ratio * text.length);
  }

  // Try finding first few words
  const words = normalizedQuote.split(' ').slice(0, 5).join(' ');
  if (words.length > 10) {
    const wordIdx = normalizedText.indexOf(words);
    if (wordIdx !== -1) {
      const ratio = wordIdx / normalizedText.length;
      return Math.floor(ratio * text.length);
    }
  }

  return -1;
}

/**
 * Get human-readable location (e.g., "paragraph 3")
 */
function getHumanReadableLocation(text: string, offset: number): string {
  const beforeOffset = text.slice(0, offset);
  const paragraphs = beforeOffset.split(/\n\n+/);
  return `Paragraph ${paragraphs.length}`;
}

/**
 * Parse JSON from Claude's response
 */
function parseJsonFromResponse(text: string, chunkId?: string): Record<string, unknown> {
  const codeBlockMatch = text.match(/```(?:json)?\s*([\s\S]*?)```/);
  if (codeBlockMatch) {
    try {
      return JSON.parse(codeBlockMatch[1].trim());
    } catch (e) {
      console.error(`JSON parse error in code block for ${chunkId || 'unknown chunk'}:`, e);
    }
  }

  const jsonMatch = text.match(/\{[\s\S]*\}/);
  if (jsonMatch) {
    try {
      return JSON.parse(jsonMatch[0]);
    } catch (e) {
      console.error(`JSON parse error for ${chunkId || 'unknown chunk'}:`, e);
      console.error('JSON snippet:', jsonMatch[0].slice(0, 200), '...', jsonMatch[0].slice(-200));
    }
  }

  console.error(`Failed to extract JSON from response for ${chunkId || 'unknown chunk'}. Response start:`, text.slice(0, 500));
  return {};
}
